circuits/types/scalar/src/add.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+use snarkvm_utilities::{FromBytes, ToBytes};
+
+impl<E: Environment> Add<Scalar<E>> for Scalar<E> {
+    type Output = Scalar<E>;
+
+    fn add(self, other: Scalar<E>) -> Self::Output {
+        self + &other
+    }
+}
+
+impl<E: Environment> Add<Scalar<E>> for &Scalar<E> {
+    type Output = Scalar<E>;
+
+    fn add(self, other: Scalar<E>) -> Self::Output {
+        self + &other
+    }
+}
+
+impl<E: Environment> Add<&Scalar<E>> for Scalar<E> {
+    type Output = Scalar<E>;
+
+    fn add(self, other: &Scalar<E>) -> Self::Output {
+        &self + other
+    }
+}
+
+impl<E: Environment> Add<&Scalar<E>> for &Scalar<E> {
+    type Output = Scalar<E>;
+
+    fn add(self, other: &Scalar<E>) -> Self::Output {
+        let mut result = self.clone();
+        result += other;
+        result
+    }
+}
+
+impl<E: Environment> AddAssign<Scalar<E>> for Scalar<E> {
+    fn add_assign(&mut self, other: Scalar<E>) {
+        *self += &other;
+    }
+}
+
+impl<E: Environment> AddAssign<&Scalar<E>> for Scalar<E> {
+    fn add_assign(&mut self, other: &Scalar<E>) {
+        // Determine the variable mode.
+        if self.is_constant() && other.is_constant() {
+            // Compute the sum and set the new constant in `self`.
+            *self = witness!(|self, other| self + other);
+        } else {
+            // Instead of adding the bits of `self` and `other` directly, the scalars are
+            // converted into a field elements, and summed, before converting back to scalars.
+            // Note: This is safe as the base field is larger than the scalar field.
+            let sum = self.to_field() + other.to_field();
```

Comment:
Is it always case that the addition of two `Scalar`s in the base field will never wrap around the base field modulus?

Reply:
Yes